lib/fetch/dataURL.js

```diff
@@ -111,73 +111,17 @@ function dataURLProcessor (dataURL) {
  * @param {boolean} excludeFragment
  */
 function URLSerializer (url, excludeFragment = false) {
-  // 1. Let output be url’s scheme and U+003A (:) concatenated.
-  let output = url.protocol
-
-  // 2. If url’s host is non-null:
-  if (url.host.length > 0) {
-    // 1. Append "//" to output.
-    output += '//'
-
-    // 2. If url includes credentials, then:
-    if (url.username.length > 0 || url.password.length > 0) {
-      // 1. Append url’s username to output.
-      output += url.username
-
-      // 2. If url’s password is not the empty string, then append U+003A (:),
-      // followed by url’s password, to output.
-      if (url.password.length > 0) {
-        output += ':' + url.password
-      }
-
-      // 3. Append U+0040 (@) to output.
-      output += '@'
-    }
-
-    // 3. Append url’s host, serialized, to output.
-    output += decodeURIComponent(url.hostname)
-
-    // 4. If url’s port is non-null, append U+003A (:) followed by url’s port,
-    // serialized, to output.
-    if (url.port.length > 0) {
-      output += ':' + url.port
-    }
-  }
-
-  // 3. If url’s host is null, url does not have an opaque path,
-  // url’s path’s size is greater than 1, and url’s path[0]
-  // is the empty string, then append U+002F (/) followed by
-  // U+002E (.) to output.
-  // Note: This prevents web+demo:/.//not-a-host/ or web+demo:/path/..//not-a-host/,
-  // when parsed and then serialized, from ending up as web+demo://not-a-host/
-  // (they end up as web+demo:/.//not-a-host/).
-  // Undici implementation note: url's path[0] can never be an
-  // empty string, so we have to slightly alter what the spec says.
-  if (
-    url.host.length === 0 &&
-    url.pathname.length > 1 &&
-    url.href.slice(url.protocol.length + 1)[0] === '.'
-  ) {
-    output += '/.'
+  if (excludeFragment) {
+    url.hash = ''
   }
 
-  // 4. Append the result of URL path serializing url to output.
-  output += url.pathname
-
-  // 5. If url’s query is non-null, append U+003F (?),
-  // followed by url’s query, to output.
-  if (url.search.length > 0) {
-    output += url.search
-  }
+  const serialized = url.toString()
 
-  // 6. If exclude fragment is false and url’s fragment is non-null,
-  // then append U+0023 (#), followed by url’s fragment, to output.
-  if (excludeFragment === false && url.hash.length > 0) {
-    output += url.hash
+  if (serialized.endsWith('?')) {
+    return serialized.slice(0, -1)
```

Comment:
actually why is this needed? Point 5 says `?`  should be appended...

Reply:
2 node-fetch tests are failing without this change.

```
  1) node-fetch
       should NOT keep `?` sign in URL when no params are given:

      AssertionError: expected 'http://localhost:60850/question?' to equal 'http://localhost:60850/question'
      + expected - actual

      -http://localhost:60850/question?
      +http://localhost:60850/question

      at /Users/yagiz/Developer/undici/test/node-fetch/main.js:1540:26
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

  2) node-fetch
       should NOT preserve the hash (#) symbol:

      AssertionError: expected 'http://localhost:60850/question?' to equal 'http://localhost:60850/question'
      + expected - actual

      -http://localhost:60850/question?
      +http://localhost:60850/question

      at /Users/yagiz/Developer/undici/test/node-fetch/main.js:1562:26
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```